File: apps/mobile/lib/services/materiaux_service.dart
import '../models/calcul_materiaux.dart';
import '../models/fiche_metrage.dart';
import '../exceptions/app_exception.dart';

/// Module 3 — calcul automatique des matériaux à partir de la surface et
/// du système choisi, avec +15% de marge de perte (elite.md §2, Module 3 ;
/// §10.5).
///
/// ⚠️ Les coefficients ci-dessous (espacement des fourrures/montants,
/// couverture d'une plaque BA13, consommation de vis/enduit au m²) sont
/// des VALEURS DE CHANTIER STANDARD, non fixées par le cahier des charges
/// (qui ne donne pas de formule). À FAIRE VALIDER avec le dirigeant avant
/// tout usage réel — même principe que les seuils de marge dans elite.md.
class MateriauxService {
  static const double margePerte = 1.15; // +15%, imposé par le cahier des charges

  // Dimensions standard d'une plaque BA13 (m).
  static const double _longueurPlaqueBa13 = 2.60;
  static const double _largeurPlaqueBa13 = 1.20;
  static double get _surfacePlaque => _longueurPlaqueBa13 * _largeurPlaqueBa13; // 3.12 m²

  static const double _longueurBarreFourrureOuMontant = 4.0; // m, barre standard
  static const double _longueurBarreRailOuCorniere = 3.0;    // m, barre standard
  static const double _espacementEntraxe = 0.60;              // m, entre fourrures/montants
  static const double _consommationVisPlacoParM2 = 25;        // vis / m² de BA13
  static const double _consommationVisAutoforeuseParM2 = 10;  // vis / m² d'ossature
  static const double _mlBandeAJointParM2 = 0.4;               // ml de bande / m²
  static const double _kgEnduitParM2 = 0.3;                    // kg / m²

  ResultatCalculMateriaux calculer({
    required double surfaceM2,
    required String systeme,
    double? perimetreM,
    Map<String, double>? prixUnitaires, // clé = nom du matériau, valeur = FCFA
  }) {
    // A1 — surface invalide.
    if (surfaceM2 <= 0) {
      throw const AppException('La surface doit être un nombre positif.');
    }
    // A2 — système non sélectionné.
    if (systeme != SystemePlatrerie.corniereFourrureBa13 &&
        systeme != SystemePlatrerie.railsMontantsBa13) {
      throw const AppException('Choisissez un système de plâtrerie.');
    }

    final perimetre = perimetreM ?? _perimetreEstime(surfaceM2);

    final nombrePlaques = _avecMarge(surfaceM2 / _surfacePlaque);
    final visPlaco = _avecMarge(surfaceM2 * _consommationVisPlacoParM2);
    final visAutoforeuses = _avecMarge(surfaceM2 * _consommationVisAutoforeuseParM2);
    final bandeAJoint = _avecMarge(surfaceM2 * _mlBandeAJointParM2);
    final enduit = _avecMarge(surfaceM2 * _kgEnduitParM2);

    final lignes = <QuantiteMateriau>[
      QuantiteMateriau(
        nomMateriau: 'Plaques BA13 (2,60 × 1,20 m)',
        quantite: nombrePlaques.ceilToDouble(),
        unite: 'plaque',
        prixUnitaire: prixUnitaires?['ba13'],
      ),
    ];

    if (systeme == SystemePlatrerie.corniereFourrureBa13) {
      // Plafond : fourrures espacées de 60 cm sur la largeur, cornière en périphérie.
      final mlFourrure = _avecMarge((surfaceM2 / _espacementEntraxe));
      final nbFourrures = (mlFourrure / _longueurBarreFourrureOuMontant).ceilToDouble();
      final nbCornieres = _avecMarge(perimetre / _longueurBarreRailOuCorniere).ceilToDouble();

      lignes.add(QuantiteMateriau(
        nomMateriau: 'Fourrures (barres de ${_longueurBarreFourrureOuMontant.toStringAsFixed(0)} m)',
        quantite: nbFourrures,
        unite: 'barre',
        prixUnitaire: prixUnitaires?['fourrure'],
      ));
      lignes.add(QuantiteMateriau(
        nomMateriau: 'Cornières (barres de ${_longueurBarreRailOuCorniere.toStringAsFixed(0)} m)',
        quantite: nbCornieres,
        unite: 'barre',
        prixUnitaire: prixUnitaires?['corniere'],
      ));
    } else {
      // Cloison : rails en haut et en bas (2 × périmètre), montants tous les 60 cm.
      final mlRails = _avecMarge(perimetre * 2);
      final nbRails = (mlRails / _longueurBarreRailOuCorniere).ceilToDouble();
      final nbMontants = (_avecMarge(perimetre / _espacementEntraxe)).ceilToDouble();

      lignes.add(QuantiteMateriau(
        nomMateriau: 'Rails (barres de ${_longueurBarreRailOuCorniere.toStringAsFixed(0)} m)',
        quantite: nbRails,
        unite: 'barre',
        prixUnitaire: prixUnitaires?['rail'],
      ));
      lignes.add(QuantiteMateriau(
        nomMateriau: 'Montants (barres de ${_longueurBarreFourrureOuMontant.toStringAsFixed(0)} m)',
        quantite: nbMontants,
        unite: 'barre',
        prixUnitaire: prixUnitaires?['montant'],
      ));
    }

    lignes.addAll([
      QuantiteMateriau(nomMateriau: 'Vis placo', quantite: visPlaco.ceilToDouble(), unite: 'unité', prixUnitaire: prixUnitaires?['vis_placo']),
      QuantiteMateriau(nomMateriau: 'Vis autoforeuses', quantite: visAutoforeuses.ceilToDouble(), unite: 'unité', prixUnitaire: prixUnitaires?['vis_autoforeuse']),
      QuantiteMateriau(nomMateriau: 'Bande à joints', quantite: bandeAJoint, unite: 'ml', prixUnitaire: prixUnitaires?['bande_joint']),
      QuantiteMateriau(nomMateriau: 'Enduit de jointoiement', quantite: enduit, unite: 'kg', prixUnitaire: prixUnitaires?['enduit']),
    ]);

    return ResultatCalculMateriaux(systeme: systeme, surfaceM2: surfaceM2, lignes: lignes);
  }

  /// À partir d'une fiche de métrage déjà saisie (Module 5), réutilise la
  /// surface nette et le périmètre déjà calculés — évite une double saisie.
  ResultatCalculMateriaux calculerDepuisFiche(FicheMetrage fiche, {Map<String, double>? prixUnitaires}) {
    if (fiche.pieces.isEmpty) {
      throw const AppException('La fiche de métrage ne contient aucune pièce.');
    }
    return calculer(
      surfaceM2: fiche.surfaceNetteTotale,
      systeme: fiche.systeme,
      perimetreM: fiche.perimetreTotal,
      prixUnitaires: prixUnitaires,
    );
  }

  double _avecMarge(double valeur) => valeur * margePerte;

  /// Estimation grossière du périmètre à partir de la surface, utilisée
  /// UNIQUEMENT si le dirigeant n'a pas de fiche de métrage détaillée
  /// (hypothèse : pièce carrée) — préférer toujours calculerDepuisFiche().
  double _perimetreEstime(double surfaceM2) => 4 * (surfaceM2 > 0 ? surfaceM2 : 0).abs().toDouble().let((s) => _racine(s));

  double _racine(double x) {
    if (x <= 0) return 0;
    double estimation = x / 2;
    for (int i = 0; i < 20; i++) {
      estimation = 0.5 * (estimation + x / estimation);
    }
    return estimation;
  }
}

/// Petit utilitaire d'extension pour garder _perimetreEstime lisible
/// (équivalent d'un `.let {}` façon Kotlin, familier si vous venez d'un
/// contexte Android/Kotlin comme mentionné dans le cahier des charges).
extension _Let<T> on T {
  R let<R>(R Function(T) f) => f(this);
}

File: apps/api/src/main/java/com/eliteplaco/api/service/MateriauxService.java
package com.eliteplaco.api.service;

import com.eliteplaco.api.dto.QuantiteMateriauDTO;
import com.eliteplaco.api.dto.ResultatMateriauxDTO;
import com.eliteplaco.api.entity.FicheMetrage;
import com.eliteplaco.api.exception.AppException;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Module 3 — calcul automatique des matériaux à partir de la surface et du
 * système choisi, avec +15% de marge de perte (elite.md §2 ; §10.5).
 *
 * ⚠️ Les coefficients ci-dessous (espacement des fourrures/montants,
 * couverture d'une plaque BA13, consommation de vis/enduit au m²) sont des
 * VALEURS DE CHANTIER STANDARD, non fixées par le cahier des charges. À
 * FAIRE VALIDER avec le dirigeant avant tout usage réel — identiques à
 * celles utilisées côté Flutter (Script 6) pour rester cohérentes.
 */
@Service
public class MateriauxService {

    private static final double MARGE_PERTE = 1.15; // +15%, imposé par le cahier des charges

    private static final double LONGUEUR_PLAQUE_BA13 = 2.60;
    private static final double LARGEUR_PLAQUE_BA13 = 1.20;
    private static final double SURFACE_PLAQUE = LONGUEUR_PLAQUE_BA13 * LARGEUR_PLAQUE_BA13; // 3.12 m²

    private static final double LONGUEUR_BARRE_FOURRURE_MONTANT = 4.0; // m
    private static final double LONGUEUR_BARRE_RAIL_CORNIERE = 3.0;    // m
    private static final double ESPACEMENT_ENTRAXE = 0.60;              // m
    private static final double VIS_PLACO_PAR_M2 = 25;
    private static final double VIS_AUTOFOREUSE_PAR_M2 = 10;
    private static final double ML_BANDE_JOINT_PAR_M2 = 0.4;
    private static final double KG_ENDUIT_PAR_M2 = 0.3;

    public ResultatMateriauxDTO calculer(double surfaceM2, FicheMetrage.SystemePlatrerie systeme,
                                          Double perimetreM, Map<String, BigDecimal> prixUnitaires) {
        if (surfaceM2 <= 0) {
            throw new AppException("La surface doit être un nombre positif.");
        }
        if (systeme == null) {
            throw new AppException("Choisissez un système de plâtrerie.");
        }
        Map<String, BigDecimal> prix = prixUnitaires == null ? Map.of() : prixUnitaires;

        double perimetre = perimetreM != null ? perimetreM : Math.sqrt(surfaceM2) * 4;

        List<QuantiteMateriauDTO> lignes = new ArrayList<>();
        lignes.add(new QuantiteMateriauDTO(
                "Plaques BA13 (2,60 × 1,20 m)",
                Math.ceil(avecMarge(surfaceM2 / SURFACE_PLAQUE)), "plaque", prix.get("ba13")));

        if (systeme == FicheMetrage.SystemePlatrerie.CORNIERE_FOURRURE_BA13) {
            double mlFourrure = avecMarge(surfaceM2 / ESPACEMENT_ENTRAXE);
            double nbFourrures = Math.ceil(mlFourrure / LONGUEUR_BARRE_FOURRURE_MONTANT);
            double nbCornieres = Math.ceil(avecMarge(perimetre / LONGUEUR_BARRE_RAIL_CORNIERE));

            lignes.add(new QuantiteMateriauDTO("Fourrures (barres de 4 m)", nbFourrures, "barre", prix.get("fourrure")));
            lignes.add(new QuantiteMateriauDTO("Cornières (barres de 3 m)", nbCornieres, "barre", prix.get("corniere")));
        } else {
            double mlRails = avecMarge(perimetre * 2);
            double nbRails = Math.ceil(mlRails / LONGUEUR_BARRE_RAIL_CORNIERE);
            double nbMontants = Math.ceil(avecMarge(perimetre / ESPACEMENT_ENTRAXE));

            lignes.add(new QuantiteMateriauDTO("Rails (barres de 3 m)", nbRails, "barre", prix.get("rail")));
            lignes.add(new QuantiteMateriauDTO("Montants (barres de 4 m)", nbMontants, "barre", prix.get("montant")));
        }

        lignes.add(new QuantiteMateriauDTO("Vis placo", Math.ceil(avecMarge(surfaceM2 * VIS_PLACO_PAR_M2)), "unité", prix.get("vis_placo")));
        lignes.add(new QuantiteMateriauDTO("Vis autoforeuses", Math.ceil(avecMarge(surfaceM2 * VIS_AUTOFOREUSE_PAR_M2)), "unité", prix.get("vis_autoforeuse")));
        lignes.add(new QuantiteMateriauDTO("Bande à joints", avecMarge(surfaceM2 * ML_BANDE_JOINT_PAR_M2), "ml", prix.get("bande_joint")));
        lignes.add(new QuantiteMateriauDTO("Enduit de jointoiement", avecMarge(surfaceM2 * KG_ENDUIT_PAR_M2), "kg", prix.get("enduit")));

        boolean incomplet = lignes.stream().anyMatch(l -> l.prixUnitaire() == null);
        BigDecimal total = incomplet ? null : lignes.stream()
                .map(QuantiteMateriauDTO::sousTotal)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        return new ResultatMateriauxDTO(systeme.name(), surfaceM2, lignes, total, incomplet);
    }

    /** Réutilise la surface nette + périmètre déjà calculés d'une fiche de métrage validée (Module 5). */
    public ResultatMateriauxDTO calculerDepuisFiche(FicheMetrage fiche, Map<String, BigDecimal> prixUnitaires) {
        if (fiche.getPieces().isEmpty()) {
            throw new AppException("La fiche de métrage ne contient aucune pièce.");
        }
        double surfaceNette = fiche.getPieces().stream()
                .mapToDouble(p -> Math.max(0, (p.getLongueur().doubleValue() * p.getLargeur().doubleValue()) - p.getSurfaceDeduction().doubleValue()))
                .sum();
        double perimetre = fiche.getPieces().stream()
                .mapToDouble(p -> 2 * (p.getLongueur().doubleValue() + p.getLargeur().doubleValue()))
                .sum();

        return calculer(surfaceNette, fiche.getSysteme(), perimetre, prixUnitaires);
    }

    private double avecMarge(double valeur) {
        return valeur * MARGE_PERTE;
    }
}
